### PR TITLE
mirror-tap test

### PR DESCRIPTION
**Pull request by qq351844**
mirror tap test

### CODE REVIEW

**Comment by qq351844** [comment]
simple_mirror : networking / default mirror to 48 port
mirror_control : rest api increase rule (protocol)

File: network_tap/mirror_control.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+        req = parser.OFPFlowStatsRequest(datapath)
+        datapath.send_msg(req)
+
+
+    @set_ev_cls(ofp_event.EventOFPPacketIn, MAIN_DISPATCHER)
+    def _packet_in_handler(self, ev):
+        pass
+
+class SimpleFirewallController(ControllerBase):
```
Comment (vicky-sunshine) [inline network_tap/mirror_control.py]: please change name

**Comment by qq351844** [comment]
1) merge simple_mirror and mirror control
2) change flow priority (default = 1, protocol = 100)
3)refresh mirror_data when add mirror rule and delete the rule

File: network_tap/simple_mirror.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+from webob import Response
+from ryu.app.wsgi import ControllerBase, WSGIApplication, route
+from ryu.ofproto import ether
+from ryu.ofproto import inet
+import ryu.app.ofctl.api
+
+import mirror_data
+from route import urls
+
+simple_mirre_instance_name = 'simple_mirror_api_app'
```
Comment (vicky-sunshine) [inline network_tap/simple_mirror.py]: `simple_mirror_instance_name`

File: network_tap/simple_mirror.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+class SimpleMirror(app_manager.RyuApp):
+    OFP_VERSIONS = [ofproto_v1_3.OFP_VERSION]
+
```
Comment (vicky-sunshine) [inline network_tap/simple_mirror.py]: should add `_CONTEXTS = {'wsgi': WSGIApplication}`

File: network_tap/simple_mirror.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        ofp_parser = datapath.ofproto_parser
+
+        req = ofp_parser.OFPSetConfig(datapath, ofp.OFPC_FRAG_NORMAL, 256)
+        datapath.send_msg(req)
+
+    @set_ev_cls(ofp_event.EventOFPPacketIn, MAIN_DISPATCHER)
+    def _packet_in_handler(self, ev):
+        pass
+
+class SimpleMirror(ControllerBase):
```
Comment (vicky-sunshine) [inline network_tap/simple_mirror.py]:
```python
class MirrorControlController(ControllerBase):
```

File: network_tap/simple_mirror.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+    def __init__(self, req, link, data, **config):
+        super(MirrorControlController, self).__init__(req, link, data, **config)
+        self.simple_mirror_spp = data[simple_mirror_instance_name]
+
+    @route('network_tap', urls.url_set_mirror_port, methods=['PUT'])
+    def mirror_rule_protocolport(self, req, **kwargs):
+        simple_firewall = self.simple_firewall_spp
```
Comment (vicky-sunshine) [inline network_tap/simple_mirror.py]:
```python
simple_mirror = self.simple_mirror_spp
```

**Comment by vicky-sunshine** [comment]
please also add `__init__.py` to `route` folder in order to import route module

**Comment by qq351844** [comment]
1. default mirror can add flow
2. rest api (add action) can get data
3. add mirror flow by protocol have some problem (match is not correct)
4. rest api (delete action)  is not tested

**Comment by qq351844** [comment]
1. Default test can add flow
2. Add flow by protocol has some problem (match field)

**Comment by qq351844** [comment]
1) Users can specific the port that they want to mirror
2) add and delete function can do it
3) mirror data add priority item to delete the flow

**Pull request merged**